src/language-markdown/parser-markdown.js

```diff
@@ -7,7 +7,8 @@ const parseFrontMatter = require("../utils/front-matter");
 const { mapAst, INLINE_NODE_WRAPPER_TYPES } = require("./utils");
 const mdx = require("./mdx");
 const remarkMath = require("remark-math");
-const htmlParser = require("../language-html/parser-html").parsers.html;
+const jsParser = require("../language-js/parser-babylon").parsers
```

Comment:
Just noticed that you added `htmlParser` in #6332 and you want to replace it with `jsParser` here. I'm wondering why you need to add those parsers given that it should already be handled by `language-markdown/embed.js`:

https://github.com/prettier/prettier/blob/23146404850011972f695fb6bc2b8113c3cffbfc/src/language-markdown/embed.js#L56-L61

Reply:
It can not handle adjacent JSX nodes as I described in the first PR.

Reply:
I guess we could simply always wrap it in a JSX fragment and add an internal option to handle this case in the babylon printer.

Something like:

```js
// language-markdown/embed.js

switch (node.type) { 
  // ...

  case "jsx": 
    return textToDoc(
      `<>${node.value}</>`,
      {parser: "__js_expression", __maybe_mdx_adjacent: true}
  ); 
  
  // ...
}
```

```js
// language-js/printer-estree.js

switch (node.type) {
  // ...
 
  case "JSXFragment":
    if (
      options.__maybe_mdx_adjacent &&
      node.children.length === 1 &&
      node.children[0].type === "JSXFragment" &&
      path.getParentNode().type === "JsExpressionRoot"
    ) {
      return outputDocWithoutOpenCloseFragment;
    }
    return originalOutputDoc;

  // ...
}
```

This way we don't need to parse the same code twice.

Reply:
Great advice, although I prefer `<$></$>` which is not considered as `html/jsx` in `mdx`, should we?

Reply:
It shouldn't matter which symbol you want to use as long as it works.

Reply:
@ikatyang I tried to implement it locally and found out that we may add a lot of mdx specific logics in babel parser, of course maybe this is related to that I'm not so familiar with the code base, but it seems not reasonable to do it to me.

And also,

```jsx
// input
<div>1</div>   123

// output by embed
<div>1</div>{" "}
123

// output by wrapper
<div>1</div> 123
```

I prefer current output for mdx and want to keep it as is, what do you think?

Reply:
Not quite sure why `123` is parsed as part of the JSX node, but the current output for MDX shouldn't be hard to implement in this way, we could use some regexes to prevent the non JSX part from being passed into the babel parser:

```js
// language-markdown/parser-markdown.js

const nonJsxPart = jsxText.match(/^[^]+<\/\w+>([^]*)$/)[1];
const jsxPart = !nonJsxPart ? jsxText : jsxPart.slice(0, -nonJsxPart.length);
```

Reply:
FYI you could put those MDX-specific logic in another file, just like what we did for Vue-specific logic.

https://github.com/prettier/prettier/blob/23146404850011972f695fb6bc2b8113c3cffbfc/src/language-js/printer-estree.js#L443-L446
https://github.com/prettier/prettier/blob/23146404850011972f695fb6bc2b8113c3cffbfc/src/language-js/html-binding.js#L7

Reply:
@ikatyang Sorry about this, but can we keep it as is temporarily? I don't have much time on it recently, I will keep the refactor soon as I'm not too busy. Of course, it would be great if you could offer the help.

Reply:
No worry, there's no time limit for this PR, you can work on it whenever you have time. Let me know if you need some help.

Reply:
> Not quite sure why 123 is parsed as part of the JSX node

This behavior is correct. It should be parsed JSX. See the interactive playground on https://mdxjs.com/:

![image](https://user-images.githubusercontent.com/94334/68838480-53a80400-06c7-11ea-8d04-726fc8e023b5.png)

Reply:
@thorn0 It's not true, see https://github.com/prettier/prettier/pull/6949#discussion_r346211287

Reply:
What exactly is not true? Is that playground broken or what?

Reply:
![image](https://user-images.githubusercontent.com/94334/68845681-77257b80-06d4-11ea-8493-92511ee48841.png)

Reply:
OK, sorry I misunderstand the code, jsx with text prefix should be considered as html, without text prefix it is jsx. That's why I was in a dilemma at https://github.com/prettier/prettier/pull/6398#discussion_r315261475 .

Reply:
@ikatyang I don't think your regex `/^[^]+<\/\w+>([^]*)$/` is always correct, I prefer parsing the jsx value by `babel`, regex is not a good idea for code parser here.

The following examples are not considered as `jsx` part in your regex.

```jsx
<></>

<li/>
```

So I think we should keep it as by using `__js_expression` here. Any further idea?

Reply:
@evilebottnawi @ikatyang @thorn0 

Please review this PR again, although I didn't change the code according to @ikatyang's advice, see my explain above.

And I'm so sorry that my first PR affects so many people, so I hope we can fix is quickly. If more refactoring are needed, let's do it later.